pkg/gadgets/trace/exec/tracer/tracer_test.go

```diff
@@ -0,0 +1,288 @@
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	utilstest "github.com/kinvolk/inspektor-gadget/pkg/gadgets/internal/test"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/trace/exec/tracer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/trace/exec/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+func TestExecTracerCreate(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	tracer, err := tracer.NewTracer(&tracer.Config{}, nil,
+		func(types.Event) {})
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+	t.Cleanup(tracer.Stop)
+}
+
+func TestExecTracerStopIdempotent(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	tracer, err := tracer.NewTracer(&tracer.Config{}, nil,
+		func(types.Event) {})
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+
+	// Check that a double stop doesn't cause issues
+	tracer.Stop()
+	tracer.Stop()
+}
+
+func createTracer(
+	t *testing.T, config *tracer.Config, callback func(types.Event),
+) *tracer.Tracer {
+	t.Helper()
+
+	tracer, err := tracer.NewTracer(config, nil, callback)
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	t.Cleanup(tracer.Stop)
+
+	return tracer
+}
+
+type test struct {
+	getTracerConfig func(info *utilstest.RunnerInfo) *tracer.Config
+	runnerConfig    *utilstest.RunnerConfig
+	generateEvent   func() (int, error)
+	validateEvent   func(t *testing.T, info *utilstest.RunnerInfo, catPid int, events []types.Event)
+}
+
+// Function to generate an event used most of the times.
+// Returns pid of executed process.
+func generateEvent() (int, error) {
+	cmd := exec.Command("/bin/cat", "/dev/null")
+	if err := cmd.Run(); err != nil {
+		return 0, fmt.Errorf("running command: %w", err)
+	}
+
+	return cmd.Process.Pid, nil
+}
+
+func TestExecTracer(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	const unprivilegedUID = int(1435)
+
+	manyArgs := []string{}
+	// 19 is DEFAULT_MAXARGS - 1 (-1 because args[0] is on the first position).
+	for i := 0; i < 19; i++ {
+		manyArgs = append(manyArgs, "/dev/null")
+	}
+
+	for name, test := range map[string]test{
+		"captures_all_events_with_no_filters_configured": {
+			getTracerConfig: func(info *utilstest.RunnerInfo) *tracer.Config {
+				return &tracer.Config{}
+			},
+			generateEvent: generateEvent,
+			validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, catPid int) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(catPid),
+					Ppid:      uint32(info.Pid),
+					UID:       uint32(info.UID),
+					MountNsID: info.MountNsID,
+					Retval:    0,
+					Comm:      "cat",
+					Args:      []string{"/bin/cat", "/dev/null"},
+				}
+			}),
+		},
+		"captures_no_events_with_filter_configured": {
```

Comment:
It would be bad if tracer would capture no events if _any_ filter is configured :wink: 
```suggestion
		"produce_no_events_when_no_matching_events_occur": {
```